### PR TITLE
Stun implementation

### PR DESCRIPTION
**Pull request by moninom1**
Description of changes:

- Make STUN serializer/deserializer standalone and separate out the APIs.
- Couple with standard C library only.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/include/stun_data_types.h [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+#define STUN_ATTRIBUTE_FINGERPRINT_LENGTH           4
+#define STUN_ATTRIBUTE_HEADER_LENGTH_OFFSET         2
+#define STUN_ATTRIBUTE_HEADER_VALUE_OFFSET          4
+#define STUN_ATTRIBUTE_ADDRESS_HEADER_LENGTH        4
+#define STUN_ERROR_CODE_PACKET_ERROR_CLASS_OFFSET   2
+#define STUN_ERROR_CODE_PACKET_ERROR_CODE_OFFSET    3
+#define STUN_ERROR_CODE_PACKET_ERROR_PHRASE_OFFSET  4
+
+/* Helper macros. */
+#define ALIGN_SIZE_TO_WORD( size )                  ( ( ( size ) + 0x3 ) & ~( 0x3 ) )
```
Comment (ActoryOu) [inline source/include/stun_data_types.h]: I'd suggest to add STUN prefix to differenciate these macros with the ones of other libraries.

File: source/include/stun_data_types.h [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+extern void ( *readUINT64 ) ( uint64_t *, uint8_t * );
+
+void init_endianness();
+
+/*-----------------------------------------------------------*/
+
+typedef enum StunResult
+{
+    STUN_RESULT_OK,
+    STUN_RESULT_BAD_PARAM,
```
Comment (ActoryOu) [inline source/include/stun_data_types.h:105]: Can we align these error return with SDP ones with different base?
Reference:
https://github.com/awslabs/amazon-kinesis-video-streams-sdp/blob/main/source/include/sdp_data_types.h#L30-L50

File: source/include/stun_data_types.h [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+#define STUN_IPV6_ADDRESS_SIZE      0x10
+
+/* STUN context flags. */
+#define STUN_FLAG_FINGERPRINT_ATTRIBUTE             ( 1 << 0 )
+#define STUN_FLAG_INTEGRITY_ATTRIBUTE               ( 1 << 1 )
+#define STUN_FLAG_USE_CANDIDATE_ATTRIBUTE           ( 1 << 2 )
+#define STUN_FLAG_DONT_FRAGMENT_ATTRIBUTE           ( 1 << 3 )
+
+/* Read/Write flags. */
+#define WRITE_UINT16 ( *writeUINT16 )
```
Comment (ActoryOu) [inline source/include/stun_data_types.h]: Same as previous comment, add STUN prefix to reduce the possibility of duplicate definition.

File: source/include/stun_data_types.h [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    STUN_ATTRIBUTE_TYPE_ICE_CONTROLLED      = 0x8029,
+    STUN_ATTRIBUTE_TYPE_ICE_CONTROLLING     = 0x802A,
+} StunAttributeType_t;
+/*-----------------------------------------------------------*/
+typedef struct StunContext
+{
+    const char * pStart;
+    size_t totalLength;
+    size_t currentIndex;
+    uint32_t flags;
```
Comment (ActoryOu) [inline source/include/stun_data_types.h]: Can we define this with more specific name? To show what flag it is representing?

File: source/include/stun_data_types.h [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+} StunContext_t;
+
+typedef struct StunHeaders
+{
+    StunMessageType_t messageType;
+    uint8_t transactionId[ STUN_HEADER_TRANSACTION_ID_LENGTH ];
```
Comment (ActoryOu) [inline source/include/stun_data_types.h:179]: Why array here? Can we use pointer to save more memory instead?

File: source/include/stun_data_types.h [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    StunAttributeType_t attributeType;
+    const uint8_t * pAttributeValue;
+    uint16_t attributeValueLength;
+} StunAttribute_t;
+
+typedef struct StunAddressAttribute
+{
+    uint16_t family;
+    uint16_t port;
+    uint8_t address[ 16 ];
```
Comment (ActoryOu) [inline source/include/stun_data_types.h:193]: Why array here? Can we use pointer to save more memory instead?

File: source/stun_deserializer.c [hunk 7]
```diff
@@ -0,0 +1,800 @@
...
+StunResult_t StunDeserializer_Init( StunContext_t * pCtx,
+                                    const uint8_t * pStunMessage,
+                                    size_t stunMessageLength,
+                                    StunHeader_t * pStunHeader )
+{
+    StunResult_t result = STUN_RESULT_OK;
+    uint32_t magicCookie;
+    uint16_t messageLengthInHeader;
+
+    init_endianness();
```
Comment (ActoryOu) [inline source/stun_deserializer.c]: Can we limit this to only run once in the beginning?

File: source/stun_serializer.c [hunk 8]
```diff
@@ -0,0 +1,862 @@
...
+        }
+
+        pCtx->currentIndex += STUN_HEADER_LENGTH;
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+StunResult_t StunSerializer_xorIpAddress( StunAttributeAddress_t *ipAddress,
```
Comment (ActoryOu) [inline source/stun_serializer.c]:
```suggestion
StunResult_t StunSerializer_XorIpAddress( StunAttributeAddress_t *ipAddress,
```
Should xor be Xor as coding convention?

**Review by ActoryOu — COMMENTED** [review]
- Add CMakeLists.txt for WebRTC SDK to include this library as external library.
    - Refer to [SDP CMakeLists.txt](https://github.com/awslabs/amazon-kinesis-video-streams-sdp/blob/main/CMakeLists.txt)
- Could you help describe how do you test on both big/little endian environments?

File: source/include/stun_data_types.h (hunk 3 above)
Comment (moninom1) [reply source/include/stun_data_types.h]: Sure, i am updating endianess code according to gaurav;s comment so will take care of this in the same.

File: source/stun_deserializer.c (hunk 7 above)
Comment (moninom1) [reply source/stun_deserializer.c]: Sure, i am updating endianess code according to gaurav's comment so will take care of this in the same.

File: source/include/stun_deserializer.h [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+                                                           uint8_t *pTransactionId );
+
+StunResult_t StunDeserializer_ParseAttributeResponseAddress( const StunContext_t * pCtx,
+                                                             const StunAttribute_t * pAttribute,
+                                                             StunAttributeAddress_t *pStunMappedAddress,
+                                                             uint8_t *pTransactionId );
+
+StunResult_t StunDeserializer_ParseAttributeSourceAddress( const StunContext_t * pCtx,
+                                                           const StunAttribute_t * pAttribute,
+                                                           StunAttributeAddress_t *pStunMappedAddress,
```
Comment (amazonKamath) [inline source/include/stun_deserializer.h:81]: NIT: space between asterix needs to be consistent across files and other components.

File: source/stun_serializer.c [hunk 10]
```diff
@@ -0,0 +1,897 @@
...
+    StunResult_t result = STUN_RESULT_OK;
+    uint16_t attributeValueLength = sizeof( uint32_t );
+
+    if( pCtx == NULL )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( ( result == STUN_RESULT_OK ) &&
+        ( pCtx->pStart != NULL ) )
```
Comment (amazonKamath) [inline source/stun_serializer.c:174]: what should be done if this is NULL ?

File: source/stun_serializer.c [hunk 11]
```diff
@@ -0,0 +1,897 @@
...
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        /* Update flags. */
+        if( attributeType == STUN_ATTRIBUTE_TYPE_FINGERPRINT )
+        {
+            pCtx->attributeFlag |= STUN_FLAG_FINGERPRINT_ATTRIBUTE;
+        }
+        if( attributeType == STUN_ATTRIBUTE_TYPE_MESSAGE_INTEGRITY )
```
Comment (amazonKamath) [inline source/stun_serializer.c:206]: else if?

File: source/stun_serializer.c [hunk 12]
```diff
@@ -0,0 +1,897 @@
...
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+static StunResult_t AddAttributeFlag( StunContext_t * pCtx,
+                                      StunAttributeType_t attributeType )
+{
+    StunResult_t result = STUN_RESULT_OK;
```
Comment (amazonKamath) [inline source/stun_serializer.c:115]: L115 to L138 seems to be similar in other function. Can we have an helper function  for these tests or is it overkill or they are not all same (I'm lazy to check)

File: source/stun_serializer.c [hunk 13]
```diff
@@ -0,0 +1,897 @@
...
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( result == STUN_RESULT_OK &&
+        pCtx->pStart != NULL &&
+        ( attributeType == STUN_ATTRIBUTE_TYPE_XOR_MAPPED_ADDRESS ||
+          attributeType == STUN_ATTRIBUTE_TYPE_XOR_RELAYED_ADDRESS ||
+          attributeType == STUN_ATTRIBUTE_TYPE_XOR_PEER_ADDRESS ))
+    {
+            StunSerializer_XorIpAddress(pCtx, pstunMappedAddress, transactionId);
```
Comment (amazonKamath) [inline source/stun_serializer.c:437]: indent?

File: source/stun_serializer.c (hunk 10 above)
Comment (moninom1) [reply source/stun_serializer.c:174]: pCtx->pStart == NULL case is to just find out the total length, so it will just update the currentIndex

**Pull request closed without merging**